NETOBSERV-1259: Adding Netobserv DNS tracking blog (#51)
netobserv/documents: blogs/dns_tracking/dns_tracking.md

# Network Observability Per Flow DNS tracking

![logo](./images/dns_tracking_logo.png)

By: Julien Pinsonneau, Mehul Modi and Mohamed S. Mahmoud

In today's interconnected digital landscape, Domain Name System (DNS) tracking
plays a crucial role in networking and security.
DNS resolution is a fundamental process that translates human-readable domain
names into IP addresses, enabling communication between devices and servers.
However, this process also presents opportunities for monitoring and analysis,
which can be achieved through innovative technologies like
eBPF (extended Berkeley Packet Filter).
In this blog post, we'll delve into the world of DNS tracking using eBPF
tracepoint hooks, exploring how this powerful combination can be used for
various purposes, including network monitoring and security enhancement.

## Understanding DNS Resolution

Before diving into the specifics of eBPF tracepoint hooks, let's briefly
recap how DNS resolution works.
When you enter a website's domain name (e.g., www.example.com) in your
browser, your computer needs to find the corresponding IP address.

> This process involves multiple steps, including querying DNS servers,
> caching responses, and ultimately obtaining the IP address for
> establishing a connection.

Reviewer: ```suggestion
This process involves multiple steps, including querying DNS servers, obtaining the IP address to establish a connection and caching response for future re-occurrence of same DNS query
```

## Utilizing Tracepoint Hooks for DNS Tracking

Tracepoint hooks are predefined points in the Linux kernel where eBPF
programs can be attached to capture and analyze specific events.
For DNS tracking, we leveraged tracepoint hooks associated with DNS
resolution processes specifically `tracepoint/net/net_dev_queue` tracepoint,
Then we parse the DNS header to determine if its query or response, attempt
to correlate query and response for specific DNS transaction and then record the

> elapsed time as well as enrich the flow with DNS header's information like DNS Id

Reviewer: ```suggestion
elapsed time to determine DNS latency as well as to enrich the flow with DNS header's information like DNS Id
```

> and DNS flags to help UI filtering.

Reviewer: ```suggestion
and DNS response codes to help filtering in Network Observability console.
```

Once `DnsTacking` feature enabled, the Console plugin will automatically adapt to provide
additionnal filters and show informations across views.

Open your OCP Console and move to 
`Administrator view` -> `Observe` -> `Network Traffic` page as usual.

Three new filters, `DNS Id`, `DNS Latency` and `DNS Response Code` will be available 
in the common section:

![dns filters](./images/dns_filters.png)

The first one will allow you to filter on a specific DNS Id to correlate with your query.

![dns id](./images/dns_id.png)

The second one helps to identify potential performance issues by looking at DNS resolution latency.

![dns latency more than](./images/dns_latency_more_than.png)

The third filter surfaces DNS response codes, which can help detect errors or unauthorized resolutions.

![dns rcode](./images/dns_response_code.png)

### Overview
New graphs will be introduced in the `advanced options` -> `manage panels` popup:

![advanced options 1](./images/advanced_options1.png)

- Top 5 average DNS latencies
- Top 5 DNS response code
- Top 5 DNS response code stacked with total

![dns graphs 1](./images/dns_graphs1.png)
![dns graphs 2](./images/dns_graphs2.png)

### Traffic flows

> The table view will get the new DNS related columns `Id`, `Latency` and `Response code` available from 

Reviewer: ```suggestion
The table view will get the new DNS columns `Id`, `Latency` and `Response code` available from 
```

the `advanced options` -> `manage columns` popup

![advanced options 2](./images/advanced_options2.png)

> The DNS related flows will show these informations in both table and side panel:

Reviewer: ```suggestion
The DNS flows will show these informations in both table and side panel:
```

section

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true

>       enableDNSTracking: true

Reviewer 2: That should end like this:
https://github.com/netobserv/network-observability-operator/pull/394/files#r1304927133

Let's keep an eye on the final output

>         - DnsTacking

Reviewer 2: That should end like this:
https://github.com/netobserv/network-observability-operator/pull/394/files#r1304927133

Let's keep an eye on the final output

```

## A quick tour in the UI

Once `DnsTacking` feature enabled, the Console plugin will automatically adapt to provide
additionnal filters and show informations across views.

Open your OCP Console and move to 
`Administrator view` -> `Observe` -> `Network Traffic` page as usual.

Three new filters, `DNS Id`, `DNS Latency` and `DNS Response Code` will be available 
in the common section:

![dns filters](./images/dns_filters.png)

The first one will allow you to filter on a specific DNS Id to correlate with your query.

![dns id](./images/dns_id.png)

The second one helps to identify potential performance issues by looking at DNS resolution latency.

![dns latency more than](./images/dns_latency_more_than.png)

> The third filter surfaces DNS response codes, which can help detect errors or unauthorized resolutions.

Reviewer: what does "unauthorized resolutions" mean? or should we simplify say "which can help detect DNS error sources"

Author: Unauthorized it’s one of the response code in the flags

![dns rcode](./images/dns_response_code.png)

### Overview
New graphs will be introduced in the `advanced options` -> `manage panels` popup:

![advanced options 1](./images/advanced_options1.png)

- Top 5 average DNS latencies
- Top 5 DNS response code
- Top 5 DNS response code stacked with total

![dns graphs 1](./images/dns_graphs1.png)
![dns graphs 2](./images/dns_graphs2.png)

### Traffic flows

> The table view will get the new DNS related columns `Id`, `Latency` and `Response code` available from 

Reviewer: probably just "DNS columns" (-related-)

the `advanced options` -> `manage columns` popup

![advanced options 2](./images/advanced_options2.png)

> The DNS related flows will show these informations in both table and side panel:

Reviewer: here. "The DNS flows"

![dns table](./images/dns_table.png)

## Future support

> - looking at adding mDNS

Reviewer: if this stems from the observations during testing, I don't think we saw anything about mDNS in our tests, eventually I realized the port 5353 we were seeing is the internal port that dns pods are using, the service is exposed via port 53

Author: No this for support tracking mDNS where need to deal with multiple responses
